File: julia/work/2017/WF/B.py
import collections
import functools
import heapq
import itertools
import math
import re
import sys
from fractions       import gcd
from fractions       import Fraction
from multiprocessing import Pool    
from operator        import itemgetter
from operator        import mul
from functools       import reduce

class myin(object) :
    def __init__(self,default_file=None,buffered=False) :
        self.fh = sys.stdin
        self.buffered = buffered
        if(len(sys.argv) >= 2) : self.fh = open(sys.argv[1])
        elif default_file is not None : self.fh = open(default_file)
        if (buffered) : self.lines = self.fh.readlines()
        self.lineno = 0
    def input(self) : 
        if (self.buffered) : ans = self.lines[self.lineno]; self.lineno += 1; return ans
        return self.fh.readline()
    def strs(self) :   return self.input().rstrip().split()
    def ints(self) :   return (int(x) for x in self.input().rstrip().split())
    def bins(self) :   return (int(x,2) for x in self.input().rstrip().split())
    def floats(self) : return (float(x) for x in self.input().rstrip().split())

def doit(fn=None,multi=False) :
    IN = myin(fn)
    t, = IN.ints()
    inputs = [ getInputs(IN) for x in range(t) ]
    if (not multi) : 
        for tt,i in enumerate(inputs,1) :
            ans = solve(i)
            printOutput(tt,ans)
    else :
        with Pool(processes=32) as pool : outputs = pool.map(solve,inputs)
        for tt,ans in enumerate(outputs,1) :
            printOutput(tt,ans)

#####################################################################################################

def printOutput(tt,ans) :
    print("Case #%d: %d %d" % (tt,ans[0],ans[1]))

def getInputs(IN) :
    S,C = IN.ints()
    O = [0] * C
    V = [0] * C
    for i in range(C) :
        aa = IN.input().rstrip().split()
        O[i] = aa[0]
        V[i] = int(aa[1])
    return (S,C,O,V)

def tadd(a,b) : return (a[0]+b*a[1],a[1])
def tsub(a,b) : return (a[0]-b*a[1],a[1])
def tmul(a,b) : return (a[0]*b,a[1])
def tdiv(a,b) : return (a[0],a[1]*b) if b > 0 else (-a[0],-a[1]*b)
def tgt(a,b) :  return b[1]*a[0] > a[1]*b[0]
def tlt(a,b) :  return b[1]*a[0] < a[1]*b[0]

def compressCards(O,V) :
    possum = 0
    negsum = 0
    posmul = 1
    posdiv = 1
    negmul = []
    negdiv = []
    zeromul = False
    for (o,v) in zip(O,V) :
        if v == 0 :
            if o == '*' :
                zeromul = True
        elif v < 0 :
            if   o == '+' : negsum += v
            elif o == '-' : possum -= v
            elif o == '*' : negmul.append(v)
            elif o == '/' : negdiv.append(v)
        else :
            if   o == '+' : possum += v
            elif o == '-' : negsum -= v
            elif o == '*' : posmul *= v
            elif o == '/' : posdiv *= v
    O2,V2 = [],[]
    if possum > 0 :
        O2.append("+"); V2.append(possum)
    if negsum < 0 :
        O2.append("+"); V2.append(negsum)
    if posmul > 1 :
        O2.append("*"); V2.append(posmul)
    if posdiv > 1 :
        O2.append("/"); V2.append(posdiv)
    if zeromul :
        O2.append("*"); V2.append(0)
    negmul.sort()
    if negmul : O2.append("*"); V2.append(negmul.pop())
    if negmul : O2.append("*"); V2.append(negmul.pop())
    if negmul : O2.append("*"); V2.append(reduce(mul,negmul,1))
    negdiv.sort()
    if negdiv : O2.append("/"); V2.append(negdiv.pop())
    if negdiv : O2.append("/"); V2.append(negdiv.pop())
    if negdiv : O2.append("/"); V2.append(reduce(mul,negdiv,1))
    return (O2,V2)

def solve(inp) :
    (S,C,O,V) = inp
    (O2,V2) = compressCards(O,V)
    C2 = len(O2)
    maxbm = 1<<C-1
    minvals = [(1,1) for i in range(1<<C2)]
    maxvals = [(1,1) for i in range(1<<C2)]
    minvals[0] = (S,1)
    maxvals[0] = (S,1)
    for i in range(1,1<<C2) :
        first = True
        for j in range(C2) :
            bm = 1 << j
            if i & bm == 0: continue
            residual = i & ~bm
            v1,v2 = (1,1),(1,1)
            if O2[j] == "+" :
                v1 = tadd(maxvals[residual],V2[j])
                v2 = tadd(minvals[residual],V2[j])
            elif O2[j] == "-" :
                v1 = tsub(maxvals[residual],V2[j])
                v2 = tsub(minvals[residual],V2[j])
            elif O2[j] == "*" :
                v1 = tmul(maxvals[residual],V2[j])
                v2 = tmul(minvals[residual],V2[j])
            elif O2[j] == "/" :
                v1 = tdiv(maxvals[residual],V2[j])
                v2 = tdiv(minvals[residual],V2[j])
            (maxv,minv) = (v1,v2) if tgt(v1,v2) else (v2,v1)
            if first :
                maxvals[i] = maxv
                minvals[i] = minv
                first = False
            else :
                maxvals[i] = maxvals[i] if tgt(maxvals[i],maxv) else maxv
                minvals[i] = minvals[i] if tlt(minvals[i],minv) else minv
    f = Fraction(maxvals[(1<<C2)-1][0],maxvals[(1<<C2)-1][1])
    return (f.numerator,f.denominator)     

#####################################################################################################
if __name__ == "__main__" :
    doit()

File: julia/work/2017/WF/B.jl
using Printf

######################################################################################################
### a) The trick here is to turn 15! orderings to 2^15 subsets and use dynamic programing.
### b) The observation is that if I have a set of prefix operations plus one "last" operation,
###    --- The maximum value is related to either the maximum or the minimum possible value of the
###        prefix operations.
### c) This leads to a rather naiive dyanmic programming solution
######################################################################################################

function tadd(a::Tuple{BigInt,BigInt},b::BigInt)
    return (a[1]+b*a[2],a[2])
end

function tsub(a::Tuple{BigInt,BigInt},b::BigInt)
    return (a[1]-b*a[2],a[2])
end

function tmul(a::Tuple{BigInt,BigInt},b::BigInt)
    return (a[1]*b,a[2])
end

function tdiv(a::Tuple{BigInt,BigInt},b::BigInt)
    return b < 0 ? (-a[1],a[2]*(-b)) : (a[1],a[2]*b)
end

function tgt(a::Tuple{BigInt,BigInt},b::Tuple{BigInt,BigInt})
    return b[2]*a[1] > a[2]*b[1]
end

function tlt(a::Tuple{BigInt,BigInt},b::Tuple{BigInt,BigInt})
    return b[2]*a[1] < a[2]*b[1]
end

function compressCards(O::Vector{Char},V::Vector{Int64})
    possum = BigInt(0)
    negsum = BigInt(0)
    posmul = BigInt(1)
    posdiv = BigInt(1)
    negmul = Vector{BigInt}()
    negdiv = Vector{BigInt}()
    zeromul = false
    for (o,v) in zip(O,V)
        if v == 0
            if o == '*'; zeromul = true; end
        elseif v < 0
            if     o == '+'; negsum += v
            elseif o == '-'; possum -= v
            elseif o == '*'; push!(negmul,BigInt(v))
            elseif o == '/'; push!(negdiv,BigInt(v))
            end
        else
            if     o == '+'; possum += v
            elseif o == '-'; negsum -= v
            elseif o == '*'; posmul *= v
            elseif o == '/'; posdiv *= v
            end
        end 
    end
    O2 = Vector{Char}()
    V2 = Vector{BigInt}()
    if possum > 0; push!(O2,'+'); push!(V2,possum); end
    if negsum < 0; push!(O2,'+'); push!(V2,negsum); end
    if posmul > 1; push!(O2,'*'); push!(V2,posmul); end
    if posdiv > 1; push!(O2,'/'); push!(V2,posdiv); end
    if zeromul;    push!(O2,'*'); push!(V2,zero(BigInt)); end
    sort!(negmul)
    sort!(negdiv)
    if !isempty(negmul); push!(O2,'*'); push!(V2,pop!(negmul)); end
    if !isempty(negmul); push!(O2,'*'); push!(V2,pop!(negmul)); end
    if !isempty(negmul); push!(O2,'*'); push!(V2,prod(negmul)); end
    if !isempty(negdiv); push!(O2,'/'); push!(V2,pop!(negdiv)); end
    if !isempty(negdiv); push!(O2,'/'); push!(V2,pop!(negdiv)); end
    if !isempty(negdiv); push!(O2,'/'); push!(V2,prod(negdiv)); end
    return O2,V2
end

function main(infn="")
    infile = (infn != "") ? open(infn,"r") : length(ARGS) > 0 ? open(ARGS[1],"r") : stdin
    tt = parse(Int64,readline(infile))
    for qq in 1:tt
        print("Case #$qq: ")
        S,C = [parse(Int64,x) for x in split(rstrip(readline(infile)))]
        O = fill('+',C)
        V = fill(0,C)
        for i in 1:C
            aa = split(rstrip(readline(infile)))
            O[i] = aa[1][1]
            V[i] = parse(Int64,aa[2])
        end
        O2,V2 = compressCards(O,V)
        C2 = length(O2)

        minvals = resize!(Vector{Tuple{BigInt,BigInt}}(),2^C2)
        maxvals = resize!(Vector{Tuple{BigInt,BigInt}}(),2^C2)
        minvals[1] = maxvals[1] = (BigInt(S),BigInt(1))
        #print("\n")
        for i in 2:2^C2
            first = true
            bitmask = i-1
            for j in 1:C2
                bm = 1 << (j-1)
                if bitmask & bm == 0; continue; end
                residual = bitmask & ~bm + 1
                v1,v2 = (BigInt(0),BigInt(1)),(BigInt(0),BigInt(1))
                if O2[j] == '+'
                    v1 = tadd(maxvals[residual],V2[j])
                    v2 = tadd(minvals[residual],V2[j])
                elseif O2[j] == '-'
                    v1 = tsub(maxvals[residual],V2[j])
                    v2 = tsub(minvals[residual],V2[j])
                elseif O2[j] == '*'
                    v1 = tmul(maxvals[residual],V2[j])
                    v2 = tmul(minvals[residual],V2[j])
                else O2[j] == '/'
                    v1 = tdiv(maxvals[residual],V2[j])
                    v2 = tdiv(minvals[residual],V2[j])
                end
                (maxv,minv) = tgt(v1,v2) ? (v1,v2) : (v2,v1)
                if first
                    maxvals[i] = maxv
                    minvals[i] = minv
                    first = false
                else
                    maxvals[i] = tgt(maxvals[i],maxv) ? maxvals[i] : maxv
                    minvals[i] = tlt(minvals[i],minv) ? minvals[i] : minv
                end
                #print("DEBUG: $i $bitmask $bm $(j-1) $(ops[j][1]) $(ops[j][2]) v1:($v1) v2:($v2) maxvals[i]:($(maxvals[i])) minvals[i]:($(minvals[i]))\n")

            end
        end
        ans = maxvals[2^C2][1]//maxvals[2^C2][2]
        print("$(numerator(ans)) $(denominator(ans))\n")
    end
end

main()
